feat: rework retrieval events

rvagg opened the pull request:
Draft for now, an attempt at reworking the event system now that autoretrieve is the only consumer. I've done away with the "status" object and have introduced typed events which are all the same but 3 have additional, specific fields: query-ask, failure & success. So for each of those, you can type cast and access the additional info.

I wanted to get this up so suggest how we could move beyond the awkwardness in #88 of hanging the `QueryResponse` in the event. But this PR isn't critical to get the autoretrieve work over the line, so I'm suggesting we merge #88 and treat this as a second stage of work.

hannahhoward commented on rep/event.go:
```diff
@@ -18,35 +16,151 @@ const (
+	return RetrievalEventSuccess{phase, payloadCid, storageProviderId, storageProviderAddr, receivedSize}
 }
+
+func (r RetrievalEventConnect) Code() Code                                    { return ConnectCode }
```
I wonder if we need to go as far as having all these interface methods instead of just public data fields. My baseline feeling is for simple data objects like events, I'm leaning towards public vars and not methods (among other things, serializes well if you need to do some reflect/bindnode madness). It's fine honestly though either way. 

In terms of constraining the type, I'm starting to lean on having a single private method interface method -- it's a total hack, but the closest I've gotten with go to a simple method of algebraic data types.

Anyway, just food for thought.

rvagg replied:
So, I actually started like that, all of the events had the same public properties - `Phase`, `StorageProviderID`, `PayloadCid`, although I left `Code()` as a method to give the interface something to describe, and it was working fine, until I had to consume the events in the event listener. See `OnRetrievalEvent()`. To get access to any of the properties I had to implement a type switch with a case for each variation and then end up repeating the ~4 main properties in each case because I couldn't get access to them generically! So I flipped over to this version where there's generic access via interface methods and the type switch only needs to happen to get access to the special properties on 3 of the events.

If interfaces did public properties, not just methods, that'd be awesome, but it seems like this is a pain point, and either way you have to decide whether your internal implementation has _tons_ of code, or your downstream user needs to have _tons_ of code. So I chose the former but :shrug:. If you have better ideas then I'm all ears.

hannahhoward replied:
fair.

hannahhoward reviewed (approved):
Given where we are headed with autoretrieve, I'm honestly in favor of this. Worth a cleanup and merge

rvagg commented:
added tests, reworked the reporter too so it's a single goroutine with a channel, not one per event, and got rid of the need for an `Id()` method on a subscriber

hannahhoward reviewed (approved):
LGTM
